Add GeoJSON format support

SoftwareMechanic opened the pull request:
added geojson support as overlay and tilesplugin

gkjohnson reviewed (commented):
This is great to see - and a really nice approach to overlaying and generating it.

Can you explain your use case a bit? Are you planning to have multiple polygons with multiple colors? One potential issue I see is that multiple colored geojson data sets will require separate layers of images which could cause a bit of an explosion of memory.

One thought I had is that the geojson could be converted to SVG and then you'd have all the flexibility of SVG scalability and CSS styling. This [geojson2svg](https://github.com/gagan-bansal/geojson2svg) package supports Geojson > SVG conversion. Who know's how the performance may be affected, though 🤔 Something to explore in the future.

_edit_ And sorry for the commit rework. I wanted to remove the dependency on #1302 so we can get this merged separately if needed.

SoftwareMechanic commented on example/geojson.js:
```diff
@@ -0,0 +1,252 @@
+
+	const geojsonOverlay = new GeoJSONTilesOverlay( {
+
+		levels: 14, // choose max zoom to rasterize overlay ( 12- 14 typical )
```
absolutely agree, I think that somehow we should maybe have access to the levels defined for the plugin the geojson is onto in order to manage that dynamically.
or hardcode an high number of levels if the prop is not provided?

SoftwareMechanic commented:
> As far as I understand, GeoJSON does not support custom CRS. The values should always be in latitude / longitude which will be universal. To that end the "GeoJSONImageSource" shouldn't need a custom CRS other than to adjust the implicit tiling scheme. Though I think that's probably not needed for this.

Once it happened that I received a .geojson file in EPSG:3003, but that was in fact an exception,  and not even the Cesium Viewer can render it by default, so I agree to treat them as latitude longitude coordinates, maybe we can just write a line in the docs or in a comment in the example about this

> For reference it looks like Mapbox provides a custom [styling system](https://docs.mapbox.com/style-spec/guides/) for things like this but it would be nice to avoid adding an overly complex styling system.

I agree, I would not add another 'rendering' layer, not for now at least. the more is simple the better it is imo.

> Unfortunately in order to render SVG in WebGL (or WebGPU afaik) you still have to rasterize the SVG data to a texture, so you will still have to deal with levels and the texture generation. But SVG is a good standard for these kinds of vector images and supports declarative styling out of the box rather than requiring a custom one. Once concern I have is that this will be slower, though, because of the complexity of the styles, but it will require some testing.

that was in fact my doubt about webGL, but I agree that using SVG is better for drawing and styling, and can lead us to have a good and modular management of the styling eventually.


> It looks like you've pulled and merged the cherry picked and rebased version of the branch I made into your local version, which has brought back all the styling changes I fixed. I can force push my version of the branch again but you'll need to pull it and overwrite your local version. Alternatively we can wait until https://github.com/NASA-AMMOS/3DTilesRendererJS/pull/1302 is merged and you can fix up the styling issues in this branch then. The diff is otherwise too complex to understand now, unfortunately. Let me know what you'd like to do!

🙈 I am sorry I messed up with GH desktop.
 you can forse push and I will overwrite my local version, my bad!

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,423 @@
+import * as THREE from 'three';
```
nit: Can we use specific imports to grab imports instead of `*`, similar to the other files? Like so:

```js
import { CanvasTexture, /* ... */ } from 'three';
```

SoftwareMechanic replied:
absolutely!

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,423 @@
+		geojson = null,
+		url = null, // URL or GeoJson object can be provided
+		tileDimension = 256,
+		bounds = [ - 180, - 90, 180, 90 ],
```
Instead of passing "bounds" into the image source it would be best to calculate the bounds of the geojson  object so it's a tight bounds (including some margin to account for stroke thickness, etc.

SoftwareMechanic replied:
for the moment in this new commit I will set the bounds like for other image sources using the crs bounds, but if using the bounds limited to the entity can optimize the process, I would like to implement this, even in #1302 , a lot of the services like geojson use only a small portion of the map to display data in there.

gkjohnson replied:
> if using the bounds limited to the entity can optimize the process, I would like to implement this

Calling `TilingScheme.setContentBounds` tells the tiling instance that any tiles that do not include that content bounds in its range is non existent and should not be loaded. This prevents unnecessary data loading and prevents render targets for overlays from being generated - saving memory and performance.

> even in #1302 , a lot of WMS services like geojson use only a small portion of the map to display data in there.

I'd added this in some of the WMSImageSource updates - the constructor takes a `contentBounds` field (which can be derived from the EX_GeographicBoundingBox in the WMS capabilities XML) and is used to set the content bounds of tiling [here](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/7fa6d531853f0b734b17f305540ee023e3cae1fa/src/three/plugins/images/sources/WMSImageSource.js#L45-L53) (with the full CRS range used as a fallback).

SoftwareMechanic replied:
this is awesome!! thank you for confirm me this

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,423 @@
+		tileDimension = 256,
+		bounds = [ - 180, - 90, 180, 90 ],
+		levels = 18,
+		projection = null, // 'EPSG:3857' or 'EPSG:4326' or null to auto- detect
```
I think we can remove "projection" - the geojson spec specifies that [all values are in the WGS84 coordinate system](https://stevage.github.io/geojson-spec/#section-4). It sounds like it used to be the case that GeoJSON allowed for CRS specifications but I don't expect this to be common and I'd prefer to ask the user to convert their geojson to the common representation before passing it to the library for the sake of simplicity.

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,423 @@
+
+			try {
+
+				const res = await this.fetchData( this.url );
+				// fetchData should return a Response-like object
+				// allow both Response and plain object (in case overlay fetch returns parsed JSON)
+				if ( res && typeof res.json === 'function' ) {
+
+					this.geojson = await res.json();
+
+				} else if ( res ) {
+
+					this.geojson = res;
+
+				}
```
What case are you imagining where `fetchData` will return pre-parsed JSON? I think it should be okay to assume this is performing a typical fetch in this case.

SoftwareMechanic replied:
I found out that it works parsing the data but also without parsing it, you can try i.e. with this public geojson used in 'url' parameter: [italy municipalities](https://github.com/openpolis/geojson-italy/blob/master/geojson/limits_IT_municipalities.geojson)
(it's a bit heavy as geojson, but I just used the first one I found)

but I think that the best thing to do is to parse it anyway, what do you think?

SoftwareMechanic replied:
now I got it, I am sorry I didn't understand what you mean (sometimes my english lacks), bear with me
that was a distraction error in the code, will be fixed!

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,423 @@
+
+				}
+
+			} catch ( e ) {
+
+				console.warn( 'GeoJSONImageSource: failed to fetch geojson from url', this.url, e );
+				// continue, but without geojson no tiles will be produced
+
+			}
```
I think we should just let this init function throw an error so it can be handled elsewhere if needed. This is what the other image sources do.

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,423 @@
+
+		}
+
+		return Promise.resolve( );
```
Since the function is marked as `asnyc` we shouldn't need to return a promise.

gkjohnson reviewed (commented):
Added some initial comments. I'll take a look at the drawing code later this week.

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,423 @@
+		); // [ minx,miny,maxx,maxy ] normalized
+
+		// compute tile extent in projection units ( degrees for 4326, meters for 3857 )
+		const tileInfo = this._tileExtentFromNormalized( boundsNormalized );
```
You can get the lat / lon for the tile using the `getTileBounds` and passing `normalized = false`.

SoftwareMechanic replied:
this is done by the next commit, I absolutely agree with keeping consistency.

gkjohnson replied:
It looks like this line can be switched to `tiling.getTileBounds( x, y, level, false, false );` now, right?

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,423 @@
+
+		// find features that ( quick bbox ) intersect tile
+		const features =
+			this.geojson && this.geojson.features ? this.geojson.features : [ ];
```
There's no guarantee that a geojson object has a root "features" object unless it's a "FeatureCollection". The GeoJSON root can otherwise be any of the GeoJSON geometry or feature types so we'll need to handle this more generically, as well.

SoftwareMechanic replied:
this should be fixed now, within '__drawFeatureOnCanvas_' method

gkjohnson replied:
Can you elaborate? It doesn't look like this has been addressed. I still see the `this.geojson && this.geojson.features ? this.geojson.features : [ ];` condition in the in the `_drawFeatureOnCanvas` means that only "featureCollection" geojson will draw. This valid GeoJSON does not work, for example:

```json
{
  "type": "LineString",
  "coordinates": [
    [ 100, 0 ],
    [ 201, 1 ],
    [ 201, 10 ],
    [ 100, 10 ],
    [ 100, 0 ]
  ]
}
```

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,423 @@
+
+			const f = features[ i ];
+			if ( ! f || ! f.geometry ) continue;
+			if ( ! this._featureIntersectsTile( f, tileInfo ) ) continue;
```
I wonder if we can cache the bounding boxes for each feature instead of recalculating it ever draw.

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,759 @@
+	}
+
+
+	async drawSVGImage( tokens, signal ) {
```
In the interest of getting this PR merged lets keep this PR focused on the canvas approach for rendering the GeoJSON. We can discuss an SVG approach in another issue but this code path is noticeably slower and I expect there are more complexities to it.

SoftwareMechanic replied:
for this commit I commented it out so we can have a reference eventually for the future, but if you prefer we can remove it

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,759 @@
+	}
+
+	// Compute pixel from projected coords and tileInfo
+	_projectedToPixel( pxProjX, pxProjY, bounds, width, height ) {
+
+		const { minX, minY } = bounds;
+		let { maxX, maxY } = bounds;
+
+		// avoid zero-range (prevents division by zero in mapLinear)
+		if ( maxX === minX ) maxX = minX + 1;
+		if ( maxY === minY ) maxY = minY + 1;
+
+		const x = MathUtils.mapLinear( pxProjX, minX, maxX, 0, width );
+		// canvas y origin is top, projection y increases north -> flip
+		const y = height - MathUtils.mapLinear( pxProjY, minY, maxY, 0, height );
+
+		// clamp to finite numbers
+		return [ isFinite( x ) ? x : 0, isFinite( y ) ? y : 0 ];
+
+	}
```
It may be worth rounding these returned values to integer values for the sake of performance. See [this MDN article on Canvas performance](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Optimizing_canvas#avoid_floating-point_coordinates_and_use_integers_instead) indicating that integer values are better to pass to the canvas functions.

SoftwareMechanic replied:
thank you for pointing that out

gkjohnson commented:
Thanks for the updates! I'll plan to make a few changes to simplify the code in some places. For now I think we can remove the SVG path, as well. We can always come back to this PR to get it in the history if needed.

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,386 @@
+
+		}
+
+		if ( minLon === Infinity ) {
+
+			return null;
+
+		}
```
A few final questions: is this a case you have run into or is this just a defensive check?

SoftwareMechanic replied:
I think this can be removed, it was a defensive check as you said

gkjohnson commented on src/three/plugins/images/sources/GeoJSONImageSource.js:
```diff
@@ -0,0 +1,386 @@
+
+		let [ minX, minY, maxX, maxY ] = bounds;
+
+		// avoid zero-range (prevents division by zero in mapLinear)
+		if ( maxX === minX ) maxX = minX + 1;
+		if ( maxY === minY ) maxY = minY + 1;
```
Likewise - can you explain this case? In order for this case to be encountered the "tile bounds" would have to have a total width of "0" in either dimension, right? I don't think it's possible for that to happen?

gkjohnson commented:
Thanks! I'm going to merge this and move it into the `examples/src` folder, for now. I think there are still some open questions to be worked out regarding multiple geojson assets, styling, etc. Once some of these things are worked out we can move it in to core!

Thank you again for all the work on this.
